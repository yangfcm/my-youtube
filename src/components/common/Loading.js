import React from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

const Loading = () => {
  return (
    <div className="text-center my-3 py-3">
      <FontAwesomeIcon icon="spinner" className="mx-2" />
      Loading...
    </div>
  );
};

export default Loading;
