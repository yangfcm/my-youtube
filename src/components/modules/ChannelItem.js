import React from "react";
import { Link } from "react-router-dom";

const ChannelItem = ({ channel }) => {
  return (
    <div className="card">
      <Link to={`/channel/${channel.snippet.resourceId.channelId}`}>
        <div className="pt-3 d-flex justify-content-center">
          <img
            src={channel.snippet.thumbnails.medium.url}
            alt={channel.title}
            className="card-img-top rounded-circle"
            style={{ width: "80%", height: "auto" }}
          />
        </div>
      </Link>
      <div className="card-body">
        <Link to={`/channel/${channel.snippet.resourceId.channelId}`}>
          <h5 className="card-title text-center">{channel.snippet.title}</h5>
        </Link>
      </div>
    </div>
  );
};

export default ChannelItem;
