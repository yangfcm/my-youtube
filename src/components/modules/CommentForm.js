import React from "react";
import { connect } from "react-redux";
import { addComment } from "../../actions/app";
import { clearError } from "../../actions/error";

class CommentForm extends React.Component {
  state = {
    comment: "",
    error: "",
    isPublishing: false,
    mode: "",
    myComments: null,
  };

  componentDidMount = () => {
    // console.log("did mount", this.props);
    if (this.props.video) {
      this.setState({
        mode: "comment",
      });
    } else if (this.props.commentId) {
      this.setState({
        mode: "reply",
      });
    } else if (this.props.comment) {
      this.setState({
        mode: "update",
        comment: this.props.comment,
      });
    }
    // console.log(this.state.mode);
  };

  addComment = async () => {
    const accessToken = localStorage.getItem("access_token");
    this.setState({
      isPublishing: true,
    });
    // console.log(this.props.video.id, this.props.video.snippet.channelId);
    // console.log("publish comment", this.state.comment.trim());
    await this.props.addComment(
      this.props.video.channelId,
      this.props.video.id,
      this.state.comment.trim(),
      accessToken
    );
    if (this.props.error) {
      this.setState({
        error: this.props.error,
        isPublishing: false,
      });
    }
    if (this.props.myComments) {
      this.setState({
        myComments: this.props.myComments,
        isPublishing: false,
        comment: "",
        error: "",
      });
    }
  };

  // replyComment = () => {};
  // updateComment = () => {};

  handlePublishComment = (e) => {
    e.preventDefault();
    if (this.state.mode === "reply") {
      // Reply a comment
    } else if (this.state.mode === "comment") {
      // Add a comment on video
      this.addComment();
    } else if (this.state.mode === "update") {
      // Update a comment/reply
    }
    this.props.clearError();
  };

  handleCancel = () => {
    this.setState({
      comment: "",
      error: "",
      isPublishing: false,
      myComments: null,
    });
  };

  render() {
    return this.props.auth.signedIn ? (
      <div>
        <form onSubmit={this.handlePublishComment}>
          <div className="form-group">
            <label>Publish your comment</label>
            <textarea
              className="form-control"
              name="comment"
              value={this.state.comment}
              onChange={(e) => {
                this.setState({
                  comment: e.target.value,
                });
              }}
            ></textarea>
          </div>
          {this.state.error && (
            <div className="alert alert-danger">{this.state.error}</div>
          )}
          <div className="d-flex justify-content-end">
            <button
              type="submit"
              className="btn btn-info mr-3"
              disabled={!this.state.comment.trim() || this.state.isPublishing}
            >
              {this.state.isPublishing ? "Publishing..." : "Comment"}
            </button>
            <button
              type="button"
              className="btn btn-secondary"
              onClick={this.handleCancel}
            >
              Cancel
            </button>
          </div>
        </form>
      </div>
    ) : (
      <div className="text-danger text-center">You are not signed in</div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    auth: state.auth,
    error: state.error,
    myComments: state.comment.myComments,
  };
};
export default connect(mapStateToProps, {
  addComment,
  clearError,
})(CommentForm);
