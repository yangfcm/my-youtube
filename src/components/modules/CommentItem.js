import React from "react";
import moment from "moment";

const CommentItem = ({ comment }) => {
  return (
    <React.Fragment>
      <div className="d-flex mb-2" style={{ overflow: "auto" }}>
        <div className="mr-3">
          <img
            src={comment.authorProfileImageUrl}
            alt={comment.authorDisplayName}
            className="rounded-circle"
          />
        </div>
        <div>
          <div className="mb-1">
            <span className="font-weight-bold mr-2">
              {comment.authorDisplayName}
            </span>
            <span className="text-muted">
              {moment(comment.publishedAt).fromNow()}
            </span>
          </div>
          <div dangerouslySetInnerHTML={{ __html: comment.textDisplay }}></div>
        </div>
      </div>
    </React.Fragment>
  );
};

export default CommentItem;
