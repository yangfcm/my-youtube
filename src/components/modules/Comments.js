import React from "react";

class Comments extends React.Component {
  render() {
    return <div>Comment</div>;
  }
}

export default Comments;
