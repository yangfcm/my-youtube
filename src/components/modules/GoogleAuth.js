import React from "react";
import { connect } from "react-redux";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { signIn, signOut } from "../../actions/auth";

class GoogleAuth extends React.Component {
  componentDidMount() {
    if (!window.gapi) {
      return;
    }
    window.gapi.load("client:auth2", () => {
      window.gapi.client
        .init({
          clientId: process.env.REACT_APP_CLIENT_ID,
          scope:
            "email https://www.googleapis.com/auth/userinfo.profile https://www.googleapis.com/auth/youtube",
        })
        .then(() => {
          this.auth = window.gapi.auth2.getAuthInstance();
          this.onAuthChange(this.auth.isSignedIn.get());
          this.auth.isSignedIn.listen(this.onAuthChange);
        });
    });
  }

  onAuthChange = (isSignedIn) => {
    if (isSignedIn) {
      this.props.signIn(this.auth.currentUser.get());
    } else {
      this.props.signOut();
    }
  };

  handleSignout = () => {
    this.auth.signOut();
  };

  handleSignin = () => {
    this.auth.signIn();
  };

  renderAuthButton() {
    // console.log(this.props.auth);
    if (this.props.auth.signedIn === null) {
      return;
    }
    if (this.props.auth.signedIn === true) {
      return (
        <li
          style={{ cursor: "pointer" }}
          className="list-group-item list-group-item-warning"
          onClick={this.handleSignout}
        >
          <FontAwesomeIcon icon="sign-out-alt" />
          {"   "}
          Sign out
        </li>
      );
    } else {
      return (
        <button className="btn btn-primary" onClick={this.handleSignin}>
          <FontAwesomeIcon icon="sign-in-alt" />
          {"   "}
          Sign In
        </button>
      );
    }
  }

  render() {
    return <div>{this.renderAuthButton()}</div>;
  }
}

const mapStateToProps = (state) => {
  return {
    auth: state.auth,
  };
};

export default connect(mapStateToProps, {
  signIn,
  signOut,
})(GoogleAuth);
