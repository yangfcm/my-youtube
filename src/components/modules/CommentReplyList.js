import React from "react";
import { connect } from "react-redux";
import { fetchCommentReplies, clearError } from "../../actions/app";
import Loading from "../common/Loading";
import CommentItem from "./CommentItem";
import MoreButton from "./MoreButton";

class CommentReplyList extends React.Component {
  state = {
    replies: null,
    showReplies: false,
    error: null,
    isLoading: false,
  };

  handleToggleShowReplies = async () => {
    // console.log("reply", this.props.comment);
    /** Show/hide replies */
    this.setState((prevState) => {
      return {
        showReplies: !prevState.showReplies,
      };
    });

    /** If there's replies, do nothing */
    if (this.state.replies) return;

    /** If no reply is loaded, load the first page of replies */
    const commentId = this.props.comment.id;
    this.setState({
      isLoading: true,
    });

    await this.props.fetchCommentReplies(commentId);
    if (this.props.error) {
      this.setState({
        error: this.props.error,
        isLoading: false,
      });
      return;
    }
    this.setState({
      replies: this.props.replies,
      isLoading: false,
    });
    // console.log(this.state.replies);
  };

  handleFetchNextPageReplies = async () => {
    // console.log(this.state.replies.nextPageToken);
    const commentId = this.props.comment.id;
    const { nextPageToken } = this.props.replies;
    await this.props.fetchCommentReplies(commentId, nextPageToken);
    if (this.props.error) return;
    this.setState((state, props) => {
      return {
        replies: {
          ...state.replies,
          items: state.replies.items.concat(props.replies.items),
          nextPageToken: props.replies.nextPageToken,
        },
      };
    });
  };

  render() {
    const { comment } = this.props;
    return (
      <React.Fragment>
        {comment.snippet.totalReplyCount > 0 && (
          <button
            className="btn btn-link"
            style={{ padding: "0" }}
            onClick={this.handleToggleShowReplies}
          >
            {this.state.showReplies ? "Hide" : "View"}{" "}
            {comment.snippet.totalReplyCount === 1
              ? "Reply"
              : `${comment.snippet.totalReplyCount} replies`}
          </button>
        )}
        {this.state.showReplies && this.state.isLoading && <Loading />}
        {this.state.showReplies && !this.state.isLoading && this.state.replies && (
          <div>
            {this.state.replies.items.map((item) => {
              return (
                <div className="my-3" key={item.id}>
                  {" "}
                  <CommentItem comment={item.snippet} />
                </div>
              );
            })}
            {this.state.replies.nextPageToken && (
              <div style={{ width: "60%", margin: "0 auto" }}>
                <MoreButton onClickMore={this.handleFetchNextPageReplies} />
              </div>
            )}
          </div>
        )}
      </React.Fragment>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    replies: state.comments.replies,
    error: state.error,
  };
};

export default connect(mapStateToProps, {
  fetchCommentReplies,
  clearError,
})(CommentReplyList);
