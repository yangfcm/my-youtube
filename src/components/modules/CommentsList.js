import React from "react";
import { connect } from "react-redux";
import { fetchComments, clearError } from "../../actions/app";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import Loading from "../common/Loading";
import ErrorMessage from "../common/ErrorMessage";
import CommentItem from "./CommentItem";
import MoreButton from "./MoreButton";

class CommentsList extends React.Component {
  state = {
    comments: null
  };
  componentDidMount = async () => {
    await this.props.fetchComments(this.props.videoId);
    if (this.props.error) {
      return;
    }
    if (this.props.comments === "commentsDisabled") {
      this.setState({
        comments: {
          items: [],
          disabled: true
        }
      });
    } else {
      this.setState({
        comments: {
          pageInfo: this.props.comments.pageInfo,
          items: this.props.comments.items,
          nextPageToken: this.props.comments.nextPageToken,
          disabled: false
        }
      });
    }
  };

  componentDidUpdate = async prevProps => {
    if (prevProps.videoId !== this.props.videoId) {
      await this.props.fetchComments(this.props.videoId);
      if (this.props.error) {
        return;
      }
      if (this.props.comments === "commentsDisabled") {
        this.setState({
          comments: {
            items: [],
            disabled: true
          }
        });
      } else {
        this.setState({
          comments: {
            pageInfo: this.props.comments.pageInfo,
            items: this.props.comments.items,
            nextPageToken: this.props.comments.nextPageToken,
            disabled: false
          }
        });
      }
    }
  };

  fetchNextPageComments = async () => {
    const { nextPageToken } = this.state.comments;
    if (!nextPageToken) {
      return;
    }
    await this.props.fetchComments(this.props.videoId, nextPageToken);
    this.setState((state, props) => {
      return {
        comments: {
          pageInfo: props.comments.pageInfo,
          items: state.comments.items.concat(props.comments.items),
          nextPageToken: props.comments.nextPageToken
        }
      };
    });
  };

  render() {
    return (
      <React.Fragment>
        {!this.props.error && !this.state.comments && <Loading />}
        {this.props.error && <ErrorMessage message={this.props.error} />}
        {!this.props.error && this.state.comments && (
          <div className="mb-3">
            <h5 className="mb-3">
              <FontAwesomeIcon icon="comments" /> Comments
            </h5>
            {this.state.comments.items.length === 0 && (
              <div className="text-muted text-center">
                {this.state.comments.disabled
                  ? "Comment is disabled"
                  : "No comment"}
              </div>
            )}
            {this.state.comments.items.map(item => {
              return <CommentItem comment={item} key={item.id} />;
            })}
            {this.state.comments && this.state.comments.nextPageToken && (
              <MoreButton onClickMore={this.fetchNextPageComments} />
            )}
          </div>
        )}
      </React.Fragment>
    );
  }
}

const mapStateToProps = state => {
  return {
    comments: state.comments,
    error: state.error
  };
};
export default connect(mapStateToProps, { fetchComments, clearError })(
  CommentsList
);
