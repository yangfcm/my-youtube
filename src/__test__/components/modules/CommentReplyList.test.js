import React from "react";
import { shallow } from "enzyme";
import { CommentReplyList } from "components/modules/CommentReplyList";
import Loading from "components/common/Loading";
import CommentItem from "components/modules/CommentItem";
import {
  commentItem,
  commentItemNoReply,
  repliesResponse,
} from "__test__/fixtures/comment";

describe("Test CommentReplyList component", () => {
  let wrapper;
  const mockFetchCommentReplies = jest.fn();

  beforeEach(() => {
    wrapper = shallow(
      <CommentReplyList
        comment={commentItem}
        fetchCommentReplies={mockFetchCommentReplies}
      />
    );
    afterEach(() => {
      wrapper.unmount();
    });
  });
  it("should render component", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should not render button if there is no reply for the comment", () => {
    const noReplyWrapper = shallow(
      <CommentReplyList comment={commentItemNoReply} />
    );
    expect(noReplyWrapper.find("button.btn-link").exists()).toBe(false);
  });

  it("should render the count and replies and no replies initially", () => {
    expect(wrapper.find("button.btn-link").text()).toContain(
      commentItem.snippet.totalReplyCount
    );
    expect(wrapper.find("button.btn-link").text()).toContain("View");
    expect(wrapper.find("div#reply-list").exists()).toBe(false);
  });

  it("should render Loading component when click 'reply' button", () => {
    expect(wrapper.find(Loading).exists()).toBe(false);
    const replyButton = wrapper.find("button.btn-link");
    replyButton.simulate("click");
    expect(wrapper.find(Loading).exists()).toBe(true);
    expect(mockFetchCommentReplies).toHaveBeenCalled();
  });

  it("should render correct number of replies", () => {
    const replyButton = wrapper.find("button.btn-link");
    replyButton.simulate("click");
    wrapper.setState({
      replies: repliesResponse,
      isLoading: false,
    });
    expect(wrapper.find("div#reply-list").exists()).toBe(true);
    expect(wrapper.find(CommentItem)).toHaveLength(
      repliesResponse.items.length
    );
    replyButton.simulate("click"); // When click again, can hide replies.
    expect(wrapper.find("div#reply-list").exists()).toBe(false);
  });
});
