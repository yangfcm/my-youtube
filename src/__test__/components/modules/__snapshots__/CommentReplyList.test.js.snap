// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test CommentReplyList component should render component 1`] = `
<Fragment>
  <button
    className="btn btn-link"
    onClick={[Function]}
    style={
      Object {
        "padding": "0",
      }
    }
  >
    View
     
    5 replies
  </button>
</Fragment>
`;
